### DISCUSSION TITLE
Hybrid Expression for relationship of relationship

### DISCUSSION THREAD

**Discussion by MattSom** [sqlalchemy/sqlalchemy#7983] [Usage Questions] [answered]
I have three models linked by a relationship chain. An `entry` has `requests` and a `partner`. So every request has a related partner by this:

```python
class Request(db.Model):
    __tablename__ = "requests"

    id = db.Column(UUID, primary_key=True, default=uuid.uuid4)

    entry_id = db.Column(
        UUID,
        db.ForeignKey("entries.id", name="requests_entry_fk"),
        nullable=False,
    )
    entry = db.relationship("Entry", back_populates="requests")


class Entry(db.Model):
    __tablename__ = "entries"

    id = db.Column(UUID, primary_key=True, default=uuid.uuid4)

    requests = db.relationship(
        "Request", back_populates="entry", lazy="dynamic"
    )
    partner_id = db.Column(
        UUID, db.ForeignKey("partners.id"), nullable=False
    )
    partner = db.relationship("Partner", back_populates="entries")


class Partner(db.Model):
    __tablename__ = "partners"

    id = db.Column(UUID, primary_key=True, default=uuid.uuid4)

    entries = db.relationship("Entry", back_populates="partner", lazy="dynamic")
```

On the `Request` model I'd like to add the `Partner` as well but since the `Entry` already relates to it I figured a simple `hybrid_property` will suffice:

```python
@hybrid_property
def partner(self):
    return self.entry.partner

@partner.expression
def partner(cls):
    return select([Entry.partner]).where(cls.entry_id == Entry.id)
```

But the `expression` does not seem to work at all: 

```python
p = Request.query.first().partner
p2 = Request.query.filter_by(Request.partner.id == p.id).first()
```

```sh
Neither 'hybrid_property' object nor 'ExprComparator' object associated with Request.partner has an attribute 'id'
```

How can I do something like that?

**Comment by zzzeek** [comment]
"partner" is returning a Select object, which doesn't have an ".id" field, and you also can't select() directly from a relationship like that because it's ambiguous how the tables would be spelled out. if you wanted this to have an .id you would return an aliased() of it and also spell out the select() completely:

```py
@partner.expression
def partner(cls):
    stmt = select([Partner]).join_from(cls, Entry, cls.entry).join(Partner, Entry.partner)
    return aliased(Partner, stmt)
```

but the above is not going to really help at all because you can't add a whole SELECT like that to the WHERE clause.  the closest you could get to that would be for the whole above statement to be a correlated subquery but then you can only return one column, and correlated subqueries also perform poorly.

in short, hybrids aren't really good for this since they only return a single column expression that can't imply a JOIN being added to the FROM clause of the enclosing query automatically, and you're not going to get good results.    You should write your Query using join() explicitly.

**Reply by MattSom** [reply]
Thanks for the answer!

One thing is not clear to me yet:

> You should write your Query using join() explicitly.

Do you mean in my `@partner.expression`? Or that I shall avoid hybrid property in itself on this model?

**Reply by zzzeek** [reply]
A good way to think of this is, in SQL.

The flask-query without any partner stuff:

     > Request.query.all()
     SELECT * FROM request

Now say you add your filter, what SQL would you like it to have?

    > p = Request.query.first().partner
    >  p2 = Request.query.filter_by(Request.partner.id == p.id).first()
    ???

The normal SQL would be:

    SELECT * FROM request JOIN entity ON <onclause> JOIN partner ON <onclause> WHERE partner.id = 7

There's not any straightforward way for the "filter" part of a Query to imply JOINs like that.  this is a, well, probably a major limitation in SQLAlchemy's approach but can likely also be seen as a good thing as it limits complexity.

So what Im saying hybrid property by itself is not good at taking a statement like :

    SELECT * FROM request

and changing it to:

    SELECT * FROM request JOIN entity ON <onclause> JOIN partner ON <onclause> WHERE partner.id = 7

because the hybrid is only used in the WHERE clause.  There's not a good way to imply the JOIN part of it.

**Reply by MattSom** [reply]
Thanks for the answer and patience again.

I've been trying to understand it more and write up some solution.

```python
m = Request.query.first()
p = m.partner

join_table = select(Request).join(Entry).subquery()
partner_id = (
    select(Entry.partner_id).where(Entry.id == m.entry_id).select_from(join_table)
).scalar_subquery()
stmt = select([Partner]).where(Partner.id == partner_id)

db.session.execute(stmt)
```
But this returns the following:

```sql
OperationalError: (pymysql.err.OperationalError) (1242, 'Subquery returns more than 1 row')
[SQL: SELECT partners.id, partners.email, partners.name
FROM partners 
WHERE partners.id = (SELECT entries.partner_id 
FROM entries, (SELECT requests.id AS id, requests.entry_id AS entry_id
FROM requests INNER JOIN entries ON entries.id = requests.entry_id) AS anon_1 
WHERE entries.id = %(id_1)s)]
[parameters: {'id_1': b'Z\xc9\xdd\x01\x96\xc9OV\xb1\xce\xad#\x06\xae[I'}]
(Background on this error at: https://sqlalche.me/e/14/e3q8)
```

Interestingly if I run `partner_id` query on its own it returns the proper id:

```python
partner_id = (
     select(Entry.partner_id).where(Entry.id == m.entry_id).select_from(join_table)
)
db.session.execute(partner_id).scalar()

> UUID('944c9be4-8b65-442a-af93-b465e8c87469')
```

So maybe `scalar_subquery` is not helping in the way I expect. 

I may be on a totally wrong path right now. What I am really curious about is how would you resolve this task yourself at this point? Is it better to store the partner on the request table as well and don't try to utilize this chain of relationships?

**Comment by CaselIT** [comment, marked as answer]
If all you want to do is to run queries like these
> ```python
> p2 = Request.query.filter_by(Request.partner.id == p.id).first()
> ```

you could use `has()` on entry like so:
```py
select(Request).where(Request.entry.has(Entry.partner_id == p.id))
# or using query
Request.query.filter_by(Request.entry.has(Entry.partner_id == p.id))
```
